core/msg.c

```diff
@@ -268,71 +288,86 @@ int msg_reply_int(msg_t *m, msg_t *reply)
     return 1;
 }
 
-int msg_try_receive(msg_t *m)
+static int _msg_avail(thread_t *thread)
 {
-    return _msg_receive(m, 0);
+    return (thread->msg_waiters.next) ||
+        (thread->msg_array && cib_avail(&thread->msg_queue));
 }
 
-int msg_receive(msg_t *m)
+static void _update_flag(thread_t *thread)
 {
-    return _msg_receive(m, 1);
+    if (_msg_avail(thread)) {
+        thread->flags |= THREAD_FLAG_MSG_WAITING;
+    }
 }
 
-static int _msg_receive(msg_t *m, int block)
+int msg_receive(msg_t *m)
 {
-    unsigned state = irq_disable();
-    DEBUG("_msg_receive: %" PRIkernel_pid ": _msg_receive.\n",
+    DEBUG("msg_receive(): %" PRIkernel_pid "\n",
           sched_active_thread->pid);
 
+    unsigned state = irq_disable();
+
     thread_t *me = (thread_t*) sched_threads[sched_active_pid];
 
+    /* clear new message flag */
+    me->flags &= ~THREAD_FLAG_MSG_WAITING;
+
     int queue_index = -1;
 
     if (me->msg_array) {
         queue_index = cib_get(&(me->msg_queue));
     }
 
-    /* no message, fail */
-    if ((!block) && ((!me->msg_waiters.first) && (queue_index == -1))) {
-        irq_restore(state);
-        return -1;
-    }
-
     if (queue_index >= 0) {
-        DEBUG("_msg_receive: %" PRIkernel_pid ": _msg_receive(): We've got a queued message.\n",
+        DEBUG("msg_receive(): %" PRIkernel_pid ": We've got a queued message.\n",
               sched_active_thread->pid);
         *m = me->msg_array[queue_index];
     }
     else {
         me->wait_data = (void *) m;
     }
 
-    priority_queue_node_t *node = priority_queue_remove_head(&(me->msg_waiters));
+    list_node_t *next = list_remove_head(&me->msg_waiters);
 
-    if (node == NULL) {
-        DEBUG("_msg_receive: %" PRIkernel_pid ": _msg_receive(): No thread in waiting list.\n",
+    if (next == NULL) {
+        DEBUG("msg_receive(): %" PRIkernel_pid ": No thread in waiting list.\n",
               sched_active_thread->pid);
 
         if (queue_index < 0) {
-            DEBUG("_msg_receive(): %" PRIkernel_pid ": No msg in queue. Going blocked.\n",
+            DEBUG("msg_receive(): %" PRIkernel_pid ": No msg in queue. Going blocked.\n",
                   sched_active_thread->pid);
             sched_set_status(me, STATUS_RECEIVE_BLOCKED);
 
             irq_restore(state);
             thread_yield_higher();
 
-            /* sender copied message */
+            /* either sender copied message and set the new message flag,
+             * or the thread was woken up for another reason (e.g., timeout) */
+            DEBUG("msg_receive(): %" PRIkernel_pid ": woke up. new message: %s\n",
+                    sched_active_thread->pid,
+                    me->flags & THREAD_FLAG_MSG_WAITING ? "yes" : "no");
+
+            if (me->flags & THREAD_FLAG_MSG_WAITING) {
+                state = irq_disable();
+                _update_flag(me);
+                irq_restore(state);
+            }
+            else {
+                return -1;
```

Comment:
Please reflect this `new` return value in the docs (https://github.com/RIOT-OS/RIOT/blob/master/core/include/msg.h#L297)